### PR TITLE
ARROW-11066: [FlightRPC][Java] Make zero-copy writes a configurable option

### PR DESCRIPTION
**Pull request by lidavidm**
A user reported that in the write path in Flight/Java, we were still copying data instead of zero-copying it to the network. An initial patch was reverted because it had a use-after-free of a Netty buffer, which crashed the JVM in a test when Netty/gRPC used Unsafe. A second patch was reverted because we realized that this optimization is not safe to enable in general: gRPC buffers data before writing it to the network, so Flight applications that wish to use the zero-copy optimization must not overwrite that buffer before gRPC writes it. This patch reworks the earlier patches such that:
1. The zero-copy read and zero-copy write are controlled by separate flags that can be toggled, and
2. The zero-copy write can be toggled by the application on a per-call basis.
(Note that the zero-copy read optimization was always enabled and working, but in earlier PRs, it shared a flag with the zero-copy write optimization.)
This PR also adds a test and updates the benchmark to opt in to the zero-copy optimization explicitly at runtime.

### CODE REVIEW

**Comment by lidavidm** [comment]
@emkornfield would you be able to re-review this at some point (no rush)? This is my third attempt at fixing the zero-copy write flag in Flight; because it requires that applications not reuse buffers after writing them, I've put it behind a runtime flag so that applications can opt-in.

File: java/flight/flight-core/src/main/java/org/apache/arrow/flight/ArrowMessage.java [hunk 1]
```diff
@@ -67,7 +67,25 @@
  */
 class ArrowMessage implements AutoCloseable {
 
-  public static final boolean FAST_PATH = true;
+  // If true, serialize Arrow data by giving gRPC a reference to the underlying Arrow buffer
```
Comment (emkornfield) [inline java/flight/flight-core/src/main/java/org/apache/arrow/flight/ArrowMessage.java:70]: nit: isn't this reverse?  giving arrow access to underlying gRPC buffer?

File: java/flight/flight-core/src/main/java/org/apache/arrow/flight/ArrowMessage.java [hunk 2]
```diff
@@ -67,7 +67,25 @@
...
+  static {
+    String zeroCopyReadFlag = System.getProperty("arrow.flight.enable_zero_copy_read");
+    if (zeroCopyReadFlag == null) {
+      zeroCopyReadFlag = System.getenv("ARROW_FLIGHT_ENABLE_ZERO_COPY_READ");
+    }
+    String zeroCopyWriteFlag = System.getProperty("arrow.flight.enable_zero_copy_write");
+    if (zeroCopyWriteFlag == null) {
+      zeroCopyWriteFlag = System.getenv("ARROW_FLIGHT_ENABLE_ZERO_COPY_WRITE");
+    }
+    ENABLE_ZERO_COPY_READ = !"false".equals(zeroCopyReadFlag);
```
Comment (emkornfield) [inline java/flight/flight-core/src/main/java/org/apache/arrow/flight/ArrowMessage.java:86]: not sure if this is for consistency with other flags, but equalsIgnoreCase would be slightly more robust.  having a static helper method here to consolidate might be useful.

File: java/flight/flight-core/src/main/java/org/apache/arrow/flight/ArrowMessage.java [hunk 3]
```diff
@@ -137,21 +156,24 @@ public ArrowMessage(FlightDescriptor descriptor, Schema schema, IpcOption option
...
   }
 
   /**
    * Create an ArrowMessage from a record batch and app metadata.
    * @param batch The record batch.
    * @param appMetadata The app metadata. May be null. Takes ownership of the buffer otherwise.
+   * @param tryZeroCopy Whether to enable the zero-copy optimization.
    */
-  public ArrowMessage(ArrowRecordBatch batch, ArrowBuf appMetadata, IpcOption option) {
+  public ArrowMessage(ArrowRecordBatch batch, ArrowBuf appMetadata, boolean tryZeroCopy, IpcOption option) {
```
Comment (emkornfield) [inline java/flight/flight-core/src/main/java/org/apache/arrow/flight/ArrowMessage.java:168]: is there a reason to not reference the static variable directly instead of pass this through?  testability?

File: java/flight/flight-core/src/test/java/org/apache/arrow/flight/TestBasicOperation.java [hunk 4]
```diff
@@ -63,6 +63,12 @@
  */
 public class TestBasicOperation {
 
+  @Test
+  public void fastPathEnabledByDefault() {
```
Comment (emkornfield) [inline java/flight/flight-core/src/test/java/org/apache/arrow/flight/TestBasicOperation.java:67]: nit. just fastPathDefaults since the two are conflicting.

File: java/flight/flight-core/src/test/java/org/apache/arrow/flight/TestBasicOperation.java [hunk 5]
```diff
@@ -354,7 +360,7 @@ public void testProtobufRecordBatchCompatibility() throws Exception {
       final VectorUnloader unloader = new VectorUnloader(root);
       root.setRowCount(0);
       final MethodDescriptor.Marshaller<ArrowMessage> marshaller = ArrowMessage.createMarshaller(allocator);
-      try (final ArrowMessage message = new ArrowMessage(unloader.getRecordBatch(), null, IpcOption.DEFAULT)) {
+      try (final ArrowMessage message = new ArrowMessage(unloader.getRecordBatch(), null, false, IpcOption.DEFAULT)) {
```
Comment (emkornfield) [inline java/flight/flight-core/src/test/java/org/apache/arrow/flight/TestBasicOperation.java:363]: nit: please add parameter comment on literals

**Comment by emkornfield** [comment]
a few nits, questions about the additional parameter otherwise LGTM.

File: java/flight/flight-core/src/main/java/org/apache/arrow/flight/ArrowMessage.java (hunk 3 above)
Comment (lidavidm) [reply java/flight/flight-core/src/main/java/org/apache/arrow/flight/ArrowMessage.java:168]: For testability, and so that it can be toggled at runtime/on a per-RPC basis instead of only globally/statically.

**Comment by emkornfield** [comment]
Hi David, this LGTM was there anything outstanding we needed to discuss?

**Comment by lidavidm** [comment]
I think we should be all good (obviously let's keep an eye out for if it regresses again though), thanks for taking a look.

**Comment by emkornfield** [comment]
+1 merging.
